Neos.Neos/Classes/AssetUsage/Projection/AssetUsageProjection.php

```diff
@@ -19,24 +19,50 @@
 use Neos\ContentRepository\Core\Feature\WorkspacePublication\Event\WorkspaceWasPartiallyPublished;
 use Neos\ContentRepository\Core\Feature\WorkspacePublication\Event\WorkspaceWasPublished;
 use Neos\ContentRepository\Core\Feature\WorkspaceRebase\Event\WorkspaceWasRebased;
-use Neos\ContentRepository\Core\Feature\NodeModification\Dto\SerializedPropertyValue;
 use Neos\ContentRepository\Core\Feature\NodeModification\Dto\SerializedPropertyValues;
 use Neos\EventStore\Model\Event;
 use Neos\EventStore\Model\Event\SequenceNumber;
+use Neos\EventStore\Model\EventEnvelope;
 use Neos\EventStore\Model\EventStream\EventStreamInterface;
 use Neos\Media\Domain\Model\ResourceBasedInterface;
 use Neos\Utility\Exception\InvalidTypeException;
 use Neos\Utility\TypeHandling;
+use Neos\Neos\AssetUsage\AssetUsageFinder;
+use Neos\EventStore\DoctrineAdapter\DoctrineCheckpointStorage;
+use Neos\ContentRepository\Core\EventStore\EventNormalizer;
+use Doctrine\DBAL\Connection;
+use Neos\EventStore\CatchUp\CatchUp;
+use Neos\ContentRepository\Core\Factory\ContentRepositoryId;
+use Neos\Media\Domain\Repository\AssetRepository;
+use Neos\Media\Domain\Model\AssetVariantInterface;
+use Neos\Media\Domain\Model\AssetInterface;
+use Neos\Neos\AssetUsage\Dto\AssetIdAndOriginalAssetId;
 
-// NOTE: as workaround, we cannot reflect this class (because of an overly eager DefaultEventToListenerMappingProvider in
-// Neos.EventSourcing - which will be refactored soon). That's why we need an extra factory for this class.
-// See Neos.ContentRepositoryRegistry/Configuration/Settings.hacks.yaml for further details.
-final class AssetUsageProjection implements ProjectionInterface // TODO IMPLEMENT
+/**
+ * @implements ProjectionInterface<AssetUsageFinder>
+ * @internal
+ */
+final class AssetUsageProjection implements ProjectionInterface
 {
+    private ?AssetUsageFinder $stateAccessor = null;
+    private AssetUsageRepository $repository;
+    private DoctrineCheckpointStorage $checkpointStorage;
+    /** @var array<string, string|null> */
+    private array $originalAssetIdMappingRuntimeCache = [];
 
     public function __construct(
-        private readonly AssetUsageRepository $repository
+        private readonly EventNormalizer $eventNormalizer,
+        private readonly AssetRepository $assetRepository,
+        ContentRepositoryId $contentRepositoryId,
+        Connection $dbal,
+        AssetUsageRepositoryFactory $assetUsageRepositoyFactory,
     ) {
+        $this->repository = $assetUsageRepositoyFactory->build($contentRepositoryId);
+        $this->checkpointStorage = new DoctrineCheckpointStorage(
+            $dbal,
+            $this->repository->getTableNamePrefix() . '_checkpoint',
+            self::class
+        );
```

Comment:
Just a thought: Instead of exposing `getTableNamePrefix()` we could let the interface implement the `CheckpointStorageInterface` (adapter for the `DoctrineCheckpointStorage` with the prefix it already knows) and then replace `$this->checkpointStorage->` by `$this->repository->` in this class.

Reply:
No, sth along the lines of
```php
final class AssetUsageRepository implements CheckpointStorageInterface
{
    private DoctrineCheckpointStorage $checkpointStorage;

    public function __construct(
        private readonly Connection $dbal,
        private readonly string $tableNamePrefix,
    ) {
        $this->checkpointStorage = new DoctrineCheckpointStorage($dbal, $tableNamePrefix . '_checkpoint', AssetUsageProjection::class);
    }

    public function acquireLock(): SequenceNumber {
        return $this->checkpointStorage->acquireLock();
    }
    public function updateAndReleaseLock(SequenceNumber $sequenceNumber): void {
        $this->checkpointStorage->updateAndReleaseLock($sequenceNumber);
    }
    public function getHighestAppliedSequenceNumber(): SequenceNumber {
        return $this->checkpointStorage->getHighestAppliedSequenceNumber();
    }

// ...
}
```

I'm not 100% sure if it's worth it and whether it is much cleaner at all, but it would at least keep the table name prefix local

Reply:
Hmm, somehow interesting to have it in the repository. But this would bind CheckpointStorage to the same storage, as the repository, which I assume is not necessary. OTOH it's bound to each other by the projection anyways. 🤷‍♂️ 

Maybe @skurfuerst has an opinion on this?

Reply:
General question: Why do we need a separate table per projection? Wouldn't is work with one table for all checkpoints?

Reply:
> General question: Why do we need a separate table per projection?

Good point! I don't remember whether this was a requirement for the table lock => https://github.com/neos/neos-development-collection/issues/4152